File: benchmarks/drbg.cc
/*
 *    Copyright 2022 Cryspen Sarl
 *
 *    Licensed under the Apache License, Version 2.0 or MIT.
 *    - http://www.apache.org/licenses/LICENSE-2.0
 *    - http://opensource.org/licenses/MIT
 */

#include "Hacl_HMAC_DRBG.h"

#include "util.h"

static bytes entropy_input = from_hex("98AD3A7FE9E908452A633AA3B9E46E7E324766");
static bytes nonce = from_hex("6DF336AB15C2BED6AC6FC793705EBBAAFE");
static bytes personalization_string =
  from_hex("AF65C375F0850AF29CC3FE6C0C1D31F4");
static bytes additional_input = from_hex("1CA8F61C");
static bytes output(128);

template<class... Args>
void
HACL_Drbg_complete(benchmark::State& state, Args&&... args)
{
  auto args_tuple = std::make_tuple(std::move(args)...);
  Spec_Hash_Definitions_hash_alg algorithm = std::get<0>(args_tuple);

  for (auto _ : state) {
    Hacl_HMAC_DRBG_state state = Hacl_HMAC_DRBG_create_in(algorithm);
    Hacl_HMAC_DRBG_instantiate(algorithm,
                               state,
                               entropy_input.size(),
                               entropy_input.data(),
                               nonce.size(),
                               nonce.data(),
                               personalization_string.size(),
                               personalization_string.data());

    Hacl_HMAC_DRBG_generate(algorithm,
                            output.data(),
                            state,
                            output.size(),
                            additional_input.size(),
                            additional_input.data());

    Hacl_HMAC_DRBG_free(algorithm, state);
  }
}

BENCHMARK_CAPTURE(HACL_Drbg_complete, sha2_256, Spec_Hash_Definitions_SHA2_256)
  ->Setup(DoSetup);
BENCHMARK_CAPTURE(HACL_Drbg_complete, sha2_384, Spec_Hash_Definitions_SHA2_384)
  ->Setup(DoSetup);
BENCHMARK_CAPTURE(HACL_Drbg_complete, sha2_512, Spec_Hash_Definitions_SHA2_512)
  ->Setup(DoSetup);
BENCHMARK_CAPTURE(HACL_Drbg_complete, sha1, Spec_Hash_Definitions_SHA1)
  ->Setup(DoSetup);

BENCHMARK_MAIN();

File: rust/src/hazmat/drbg.rs
use hacl_star_sys::{
    Hacl_HMAC_DRBG_create_in, Hacl_HMAC_DRBG_free, Hacl_HMAC_DRBG_generate,
    Hacl_HMAC_DRBG_instantiate, Hacl_HMAC_DRBG_reseed, Hacl_HMAC_DRBG_state,
    Spec_Hash_Definitions_SHA1, Spec_Hash_Definitions_SHA2_256, Spec_Hash_Definitions_SHA2_384,
    Spec_Hash_Definitions_SHA2_512,
};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(u8)]
pub enum Algorithm {
    Sha1 = Spec_Hash_Definitions_SHA1 as u8,
    Sha2_256 = Spec_Hash_Definitions_SHA2_256 as u8,
    Sha2_384 = Spec_Hash_Definitions_SHA2_384 as u8,
    Sha2_512 = Spec_Hash_Definitions_SHA2_512 as u8,
}

pub enum Error {
    /// Unable to generate the requested randomness.
    UnableToGenerate,
}

pub struct Drbg {
    state: Hacl_HMAC_DRBG_state,
    alg: Algorithm,
}

impl Drbg {
    /// Create a new DRBG state with the given hash function.
    /// This also initializes the DRBG state with the given entropy, nonce and
    /// personalization string.
    pub fn new(alg: Algorithm, entropy: &[u8], nonce: &[u8], personalization: &str) -> Self {
        let state = unsafe { Hacl_HMAC_DRBG_create_in(alg as u8) };
        unsafe {
            Hacl_HMAC_DRBG_instantiate(
                alg as u8,
                state,
                entropy.len().try_into().unwrap(),
                entropy.as_ptr() as _,
                nonce.len().try_into().unwrap(),
                nonce.as_ptr() as _,
                personalization.len().try_into().unwrap(),
                personalization.as_bytes().as_ptr() as _,
            );
        }
        Self { state, alg }
    }

    /// Reseed the DRBG state.
    ///
    /// It is very unlikely that you will need this function.
    pub fn reseed(&mut self, entropy: &[u8], additional_input: &[u8]) {
        unsafe {
            Hacl_HMAC_DRBG_reseed(
                self.alg as u8,
                self.state,
                entropy.len().try_into().unwrap(),
                entropy.as_ptr() as _,
                additional_input.len().try_into().unwrap(),
                additional_input.as_ptr() as _,
            );
        }
    }

    /// Generate random bytes.
    pub fn generate(&mut self, output: &mut [u8], additional_input: &[u8]) -> Result<(), Error> {
        if unsafe {
            Hacl_HMAC_DRBG_generate(
                self.alg as u8,
                output.as_mut_ptr(),
                self.state,
                output.len().try_into().unwrap(),
                additional_input.len().try_into().unwrap(),
                additional_input.as_ptr() as _,
            )
        } {
            Ok(())
        } else {
            Err(Error::UnableToGenerate)
        }
    }
}

impl Drop for Drbg {
    fn drop(&mut self) {
        unsafe { Hacl_HMAC_DRBG_free(self.alg as u8, self.state) };
    }
}
